src/player.cpp

```diff
@@ -4599,6 +4599,73 @@ void player::on_hurt( Creature *source, bool disturb /*= true*/ )
     }
 }
 
+bool player::immune_to( body_part bp, damage_unit dam ) const
+{
+    // We use the worst-case scenario instead of random numbers
+    if( has_trait( "DEBUG_NODMG" ) || is_immune_damage( dam.type ) ) {
+        return true;
+    }
+    if( dam.type == DT_HEAT ) {
+        return false; // No one is immune to fire
+    }
+
+    // Bionics reduce the damage for bashing, cutting, and stabbing
+    if( has_active_bionic( "bio_ads" ) && power_level >= 25 ) {
+        if( dam.type == DT_BASH || dam.type == DT_CUT || dam.type == DT_STAB ) {
+            dam.amount -= 1;
+        }
+    }
+    if( has_bionic( "bio_carbon" ) ) {
+        if( dam.type == DT_BASH ) {
+            dam.amount -= 2;
+        } else if( dam.type == DT_CUT ) {
+            dam.amount -= 4;
+        } else if( dam.type == DT_STAB ) {
+            dam.amount -= 3.2;
+        }
+    }
+    //all the other bionic armors reduce bash/cut/stab by 3/3/2.4
+    static const std::vector< std::pair< std::string, std::set< body_part > > > armor_bionics = {
```

Comment:
It would be better if the order was inverted here. That is, body parts mapping to sets of bionics that protect them.